### PR TITLE
[MARCHETYPES-70] enable generating Java 9+ projects and add support for JUnit 5.x

### PR DESCRIPTION
**Pull request by JurrianFahner**
@behrangsa has done nice work, the pull request #3 was stale since more than a year ago. 

I fixed the merge conflicts of that PR. 

Is it now possible to merge? After merge pull request #3 can be closed.

### CODE REVIEW

**Comment by JurrianFahner** [comment]
@eolivelli Could you review this PR?

**Comment by eolivelli** [comment]
sure, sorry for the delay @JurrianFahner

File: maven-archetype-quickstart/src/main/resources-filtered/archetype-resources/pom.xml [hunk 1]
```diff
@@ -14,17 +44,19 @@
...
 
   <dependencies>
-    <dependency>
-      <groupId>junit</groupId>
-      <artifactId>junit</artifactId>
-      <version>4.11</version>
-      <scope>test</scope>
-    </dependency>
+#if ( ${junitVersion} == $null )
+#junit( "4.12" )
```
Comment (eolivelli) [inline maven-archetype-quickstart/src/main/resources-filtered/archetype-resources/pom.xml:56]: latest version is 4.13.1 ?
https://mvnrepository.com/artifact/junit/junit/4.13.1

**Review by eolivelli — CHANGES_REQUESTED** [review]

**Comment by JurrianFahner** [comment]
> 
> 
> sure, sorry for the delay @JurrianFahner

Really not a problem... Your feedback is appreciated!

File: maven-archetype-quickstart/src/main/resources-filtered/archetype-resources/pom.xml [hunk 2]
```diff
@@ -1,5 +1,35 @@
 <?xml version="1.0" encoding="UTF-8"?>
-
+#macro( compilerProperties $ver )
+#if ($ver == "9" || $ver == "10" || $ver == "11" || $ver == "12" || $ver == "13" || $ver == "14")
```
[the remark this answers is not in the document]
Comment (JurrianFahner) [reply maven-archetype-quickstart/src/main/resources-filtered/archetype-resources/pom.xml]: We now have support up to at least "1000", for now more than enough...

File: maven-archetype-quickstart/src/main/resources-filtered/archetype-resources/pom.xml (hunk 1 above)
Comment (JurrianFahner) [reply maven-archetype-quickstart/src/main/resources-filtered/archetype-resources/pom.xml]: I've updated the version accordingly.

**Comment by JurrianFahner** [comment]
@eolivelli I have applied your PR feedback!

**Comment by eolivelli** [comment]
last step:
- please create a JIRA here
https://issues.apache.org/jira/projects/ARCHETYPE/issues/ARCHETYPE-528?filter=allopenissues
- squash your commits into one
- in the commit message cite the JIRA ID, use "git log" in order to see how we are creating git commit messages

Here you can find a guide
https://maven.apache.org/guides/development/guide-maven-development.html#Creating_and_submitting_a_patch

thank you very much !

once we are okay with the bureaucratic stuff I will push your branch to the ASF repository, ASF jenkins will validate it and if all tests pass then I will merge to master branch.

It looks a long process, but in practice it is very easy and quick once we find our way.
We need a JIRA in order to create automatically release notes and keep track of the work, we use Jenkins in order to test your patch against a matrix of Maven Versions x Java Versions x Operating Systems

**Comment by JurrianFahner** [comment]
@eolivelli Thanks for your introduction into creating patches, that was very helpful. It is my first patch for maven anyway. Please let me know whether it complies with the maven's guidelines.

**Comment by eolivelli** [comment]
I have pushed your patch to a branch named MARCHETYPES-70

Jenkins should start the validation here:
https://ci-builds.apache.org/job/Maven/job/maven-box/job/maven-archetypes/

but I see that there is a large queue, let's wait

**Comment by eolivelli** [comment]
https://ci-builds.apache.org/job/Maven/job/maven-box/job/maven-archetypes/job/MARCHETYPES-70/

**Comment by eolivelli** [comment]
try again
https://ci-builds.apache.org/job/Maven/job/maven-box/job/maven-archetypes/job/MARCHETYPES-70/2/

**Comment by JurrianFahner** [comment]
@eolivelli The build fails with the following message:

```
----- withMaven Wrapper script -----

The JAVA_HOME environment variable is not defined correctly 

This environment variable is needed to run this program 

NB: JAVA_HOME should point to a JDK not a JRE 

script returned exit code 1
```

I believe the JAVA_HOME should be set properly, that's why the build fails.

**Comment by eolivelli** [comment]
It looks like the build passes on master branch
https://ci-builds.apache.org/job/Maven/job/maven-box/job/maven-archetypes/job/master/

We should compare the logs.
Do you have time to debug @JurrianFahner  ?

I will rerun the pipiline, in order to see if it was a transient error

**Comment by JurrianFahner** [comment]
@eolivelli I've dived into the logs, and my conclusion is the same. The buildserver seems to be broken on the branch build. In some weird way the JAVA_HOME is not set, this env property is vital to run maven anyway. 

I've also verified the build on my windows machine with java 15, it was running fine. 

How can I dive in the build specifics to understand why this JAVA_HOME is not set (the other stages were build fine)?

**Comment by JurrianFahner** [comment]
@rfscholte I saw your name mentioned in jenkins, so I assumed you know how the jenkins build server has been configured. I don't have access to the build properties, is there a repo so I can look why the JAVA_HOME couldn't be resolved on the branch for the test on windows-jdk15? See here for more information: https://ci-builds.apache.org/blue/organizations/jenkins/Maven%2Fmaven-box%2Fmaven-archetypes/detail/MARCHETYPES-70/3/pipeline

**Comment by JurrianFahner** [comment]
@eolivelli How are we going to proceed? We now know the cause of the build failure. But how are we able to fix it? I believe I don't have access to the build procedure.

**Comment by eolivelli** [comment]
it looks like we have a green build !
https://ci-builds.apache.org/blue/organizations/jenkins/Maven%2Fmaven-box%2Fmaven-archetypes/detail/MARCHETYPES-70/4/pipeline

merging this change now

**Comment by eolivelli** [comment]
thank you very much for your contribution @JurrianFahner and also for your patience.
next step is to make a release.

can you please ask on dev@maven.apache.org if someone can cut a release for maven-archetypes.
I am sorry but I cannot do it in the short term, I am busy with other priorities

**Pull request merged**